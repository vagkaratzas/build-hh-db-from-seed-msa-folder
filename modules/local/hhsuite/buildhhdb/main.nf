process HHSUITE_BUILDHHDB {
    tag "$meta.id"
    label 'process_high'

    container "quay.io/microbiome-informatics/hh-suite-db-builder:1.0.0"

    input:
    tuple val(meta), path(a3m, stageAs: 'tmp_aln/*')

    output:
    tuple val(meta), path("${prefix}"), emit: hh_db
    path "versions.yml"               , emit: 'versions'

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    ffindex_build -s ${prefix}_a3m.ff{data,index} tmp_aln
    mpirun -np ${task.cpus} ffindex_apply_mpi ${prefix}_a3m.ff{data,index} -i ${prefix}_hhm.ffindex -d ${prefix}_hhm.ffdata -- hhmake -i stdin -o stdout -v 0
    mpirun -np ${task.cpus} cstranslate_mpi -f -x 0.3 -c 4 -I a3m -i ${prefix}_a3m -o ${prefix}_cs219
    sort -k3 -n -r ${prefix}_cs219.ffindex | cut -f1 > sorting.dat
    ffindex_order sorting.dat ${prefix}_hhm.ff{data,index} ${prefix}_hhm_ordered.ff{data,index}
    mv ${prefix}_hhm_ordered.ffindex ${prefix}_hhm.ffindex
    mv ${prefix}_hhm_ordered.ffdata ${prefix}_hhm.ffdata
    ffindex_order sorting.dat ${prefix}_a3m.ff{data,index} ${prefix}_a3m_ordered.ff{data,index}
    mv ${prefix}_a3m_ordered.ffindex ${prefix}_a3m.ffindex
    mv ${prefix}_a3m_ordered.ffdata ${prefix}_a3m.ffdata

    mkdir -p ${prefix}
    mv ${prefix}_* ${prefix}/

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hh-suite: \$(hhblits -h | grep 'HHblits' | sed -n -e 's/.*\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*/\\1/p')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p ${prefix}
    touch ${prefix}/${prefix}_a3m_a3m.ffdata
    touch ${prefix}/${prefix}_a3m_a3m.ffindex
    touch ${prefix}/${prefix}_a3m_cs219.ffdata
    touch ${prefix}/${prefix}_a3m_cs219.ffindex
    touch ${prefix}/${prefix}_a3m_cs219.log1
    touch ${prefix}/${prefix}_a3m_hmm.ffdata
    touch ${prefix}/${prefix}_a3m_hmm.ffindex

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hh-suite: \$(hhblits -h | grep 'HHblits' | sed -n -e 's/.*\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*/\\1/p')
    END_VERSIONS
    """
}
