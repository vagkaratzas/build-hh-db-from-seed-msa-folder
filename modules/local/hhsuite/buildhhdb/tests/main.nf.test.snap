{
    "test_a3m_cs219.ffindex": {
        "content": [
            "test_cs219.ffindex:md5,1893a6d1cf11667bbb65520f5e8d6497"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:35.440980852"
    },
    "test_a3m_cs219.ffdata": {
        "content": [
            "test_cs219.ffdata:md5,0a03ab702bc958b26fba34f8d1c5fce9"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:35.433484897"
    },
    "test_a3m_a3m.ffdata": {
        "content": [
            "test_a3m.ffdata:md5,1d3e01ef4418ae118cce3e43cd5669aa"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:35.416143459"
    },
    "a3m - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_a3m_a3m.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_a3m.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.log1:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,ab0aa7b758ce2f2f171fbf1e14339a52"
                ],
                "hh_db": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_a3m_a3m.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_a3m.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.log1:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ab0aa7b758ce2f2f171fbf1e14339a52"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:40.471299258"
    },
    "test_a3m_a3m.ffindex": {
        "content": [
            "test_a3m.ffindex:md5,34f597e4cb2d7d3afa4c655146ff3eb0"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:35.42674239"
    },
    "test_a3m_hmm.ffindex": {
        "content": [
            "test_cs219.log.3:md5,d41d8cd98f00b204e9800998ecf8427e"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-19T13:53:35.450864683"
    }
}